### PR TITLE
357 docker compose

### LINKED ISSUE #357: create a Docker Compose file to run the application
Create an initial docker compose file, it should include the following services

app (node/express server)
db (mongo)

### PR DESCRIPTION
**Pull request by Elkrival**
This pr adds the docker compose template to run dpdash for local development

### CODE REVIEW

File: DockerSetup.md [hunk 1]
```diff
@@ -0,0 +1,78 @@
...
+
+This document outlines the set up for the development environment using docker
+
+## Requirements
+
+- [Complete all the steps up to the initialize step](./README-PNL.md#initialize)
+
+## Prepare the Configs:
+
+- Copy the contents from the dashboard directory.
```
Comment (mikestone14) [inline DockerSetup.md]: where should the contents be pasted?

File: DockerSetup.md [hunk 2]
```diff
@@ -0,0 +1,78 @@
...
+- Create a directory called configs inside the server directory of the project.
+- Open the basePathConfig.js and replace it's contents with this.
```
Comment (mikestone14) [inline DockerSetup.md]: what is the "this" that the contents are being replaced with?

File: DockerSetup.md [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+
+```js
+const basePathConfig = ''
+
+export default basePathConfig
+```
+
+- Comment lines 32-35. It should look like this.
```
Comment (mikestone14) [inline DockerSetup.md]: what file are these lines in?

File: DockerSetup.md [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+// config.database.mongo.server.sslCert = cert;
+// config.database.mongo.server.sslKey = key;
+// config.database.mongo.server.sslValidate = false;
+config.database.mongo.server.reconnectTries = 1
+config.database.mongo.server.useNewUrlParser = true
+```
+
+# Create Admin Users
+
+[The init.sh script](./README-PNL.md#initialize) outputs a dashboard directory. The directory should be in this path ${state}/dpdash/configs/dashboard.
```
Comment (mikestone14) [inline DockerSetup.md]: this links to the initialize section of the readme. Is that intentional? It is referencing the `init.sh` script which is a little confusing.

File: DockerSetup.md [hunk 5]
```diff
@@ -0,0 +1,78 @@
...
+
+## Mongo User
+
+Copy the mongo.username and mongo.password values from the /server/configs/config.js file.
```
Comment (mikestone14) [inline DockerSetup.md]: where should they be pasted?

File: DockerSetup.md [hunk 6]
```diff
@@ -0,0 +1,78 @@
...
+db.createUser(
+   {
+     user: "dpdash",
+     pwd: "Some Value"
+     roles: [ "root" ]
+   }
+)
+```
+
+# Create docker-compose file and update the following values
```
Comment (mikestone14) [inline DockerSetup.md]: why do we need to create the docker-compose file? Can we not include one in the repo?

File: DockerSetup.md [hunk 7]
```diff
@@ -0,0 +1,78 @@
+# Local DP Dash development on Docker
+
+This document outlines the set up for the development environment using docker
+
+## Requirements
+
+- [Complete all the steps up to the initialize step](./README-PNL.md#initialize)
```
Comment (mikestone14) [inline DockerSetup.md]: the initial steps do things like setting up singularity. Do we still need to do that? My hunch/hope is that we can greatly reduce the setup complexity with this change.

**Comment by Elkrival** [comment]
I removed this README file, from our conversation earlier to use env variables.

File: DockerSetup.md (hunk 7 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: DockerSetup.md (hunk 6 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: DockerSetup.md (hunk 5 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: DockerSetup.md (hunk 4 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: DockerSetup.md (hunk 3 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: DockerSetup.md (hunk 2 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: DockerSetup.md (hunk 1 above)
Comment (Elkrival) [reply DockerSetup.md]: https://github.com/AMP-SCZ/dpdash/pull/358#issuecomment-1669999988

File: server/app.js [hunk 8]
```diff
@@ -109,13 +108,13 @@ app.use(methodOverride())
...
 let mongodb
-const mongoURI = getMongoURI({ settings: config.database.mongo })
+const mongoURI = `mongodb://${process.env.MONGODB_ADMIN_USER}:${process.env.MONGODB_ADMIN_PASSWORD}@${process.env.MONGODB_HOST}:${process.env.MONGODB_PORT}/${process.env.MONGODB_CONFIG_DB}?authSource=admin`
 const mongodbPromise = co(function* () {
   return yield MongoClient.connect(mongoURI, config.database.mongo.server)
 }).then(function (res) {
   mongodb = res.db()
   app.locals.appDb = res.db()
-  app.locals.dataDb = res.db(config.database.mongo.dataDB)
+  app.locals.dataDb = res.db(process.env.MONGODB_DATA_DB)
```
Comment (mikestone14) [inline server/app.js]:
```suggestion
  app.locals.dataDb = res.db(process.env.DATA_DB_NAME)
```

File: server/app.js [hunk 9]
```diff
@@ -109,13 +108,13 @@ app.use(methodOverride())
 
 /* database setup */
 let mongodb
-const mongoURI = getMongoURI({ settings: config.database.mongo })
+const mongoURI = `mongodb://${process.env.MONGODB_ADMIN_USER}:${process.env.MONGODB_ADMIN_PASSWORD}@${process.env.MONGODB_HOST}:${process.env.MONGODB_PORT}/${process.env.MONGODB_CONFIG_DB}?authSource=admin`
```
Comment (mikestone14) [inline server/app.js]:
```suggestion
const mongoURI = process.env.MONGODB_URI
```

**Review by mikestone14 — COMMENTED** [review]
let's update the README to use the new setup instructions

File: server/app.js [hunk 10]
```diff
@@ -109,13 +108,13 @@ app.use(methodOverride())
...
 let mongodb
-const mongoURI = getMongoURI({ settings: config.database.mongo })
+const mongoURI = process.env.MONGODB_URI
 const mongodbPromise = co(function* () {
   return yield MongoClient.connect(mongoURI, config.database.mongo.server)
 }).then(function (res) {
   mongodb = res.db()
   app.locals.appDb = res.db()
-  app.locals.dataDb = res.db(config.database.mongo.dataDB)
+  app.locals.dataDb = res.db(process.env.DATA_DB_NAME)
```
Comment (mikestone14) [inline server/app.js]: I'm actually wondering if we need this as a env var or if we can hard-code it to `dpdata`. What do you think?

Comment (Elkrival) [reply server/app.js:117]: @mikestone14 [2693cc2](https://github.com/AMP-SCZ/dpdash/pull/358/commits/2693cc23cae87aa48c4c47d52963d94881487249)
this is the README Commit

**Comment by Elkrival** [comment]
@mikestone14 [2693cc2](https://github.com/AMP-SCZ/dpdash/pull/358/commits/2693cc23cae87aa48c4c47d52963d94881487249)
this is the README Commit

File: server/app.js (hunk 10 above)
Comment (Elkrival) [reply server/app.js]: yeah we can take it out

File: .env.sample [hunk 11]
```diff
@@ -0,0 +1,8 @@
+# SERVER_PORT=
+# MONGODB_PORT=
+# MONGODB_ADMIN_USER=
+# MONGODB_ADMIN_PASSWORD=
+# MONGODB_HOST=
+# MONGODB_CONFIG_DB=
```
Comment (mikestone14) [inline .env.sample]: looks like we can remove these.

File: server/bin/www.js [hunk 12]
```diff
@@ -18,7 +18,7 @@ import config from '../configs/config'
  * Get port from environment and store in Express.
  */
 
-const port = normalizePort(config.app.port)
+const port = process.env.SERVER_PORT
```
Comment (mikestone14) [inline server/bin/www.js]: should we default the port to something like 3000 in case the variable isn't set?

**Review by mikestone14 — APPROVED** [review]

**Pull request merged**